compiler/rustc_trait_selection/src/solve/eval_ctxt.rs

```diff
@@ -156,17 +163,31 @@ impl<'tcx> InferCtxtEvalExt<'tcx> for InferCtxt<'tcx> {
             var_values: CanonicalVarValues::dummy(),
             nested_goals: NestedGoals::new(),
             tainted: Ok(()),
-            inspect: (self.tcx.sess.opts.unstable_opts.dump_solver_proof_tree
-                || matches!(generate_proof_tree, GenerateProofTree::Yes))
-            .then(ProofTreeBuilder::new_root)
-            .unwrap_or_else(ProofTreeBuilder::new_noop),
+            inspect: {
+                if self.tcx.sess.opts.unstable_opts.dump_solver_proof_tree
```

Comment:
Can you pull this out of the struct constructor? It's kinda weird to have this side-effect-ful code hoisted up here.